vermouth/graph_utils.py

```diff
@@ -43,88 +38,91 @@ def add_element_attr(molecule):
+                for attr in g1_edge_keys | g2_edge_keys:
+                    attrs[attr] = (graph1.edges[graph1_nodes].get(attr, None),
+                                   graph2.edges[graph1_nodes].get(attr, None))
+            product.add_edge((graph1_node1, graph2_node1), (graph1_node2, graph2_node2), **attrs)
+    return product
+
+
+def categorical_maximum_common_subgraph(graph1, graph2, attributes=tuple()):
+    product = categorical_modular_product(graph1, graph2, attributes)
+    cliques = nx.find_cliques(product)
     # cliques is an iterator which will return a *lot* of items. So make sure
     # we never turn it into a full list.
     largest = maxes(cliques, key=len)
     matches = [dict(clique) for clique in largest]
     return matches
 
 
-def maximum_common_subgraph(reference, found, attributes=tuple()):
-    print([key for key, value in reference.nodes.items() if 'element' not in value])
-    print(list(reference.nodes.items()))
-    G = reference
-    H = found
-    P = nx.Graph()
+def maximum_common_subgraph(graph1, graph2, attributes=tuple()):
+    product = nx.Graph()
     # First, find the MCS between all nodes of degree != 1, such as the carbons
     # Nothing new or exciting here.
-    for u, v in itertools.product(G, H):
-        if all(G.node[u][attr] == H.node[v][attr] for attr in attributes):
-            if G.degree(u) != 1 and H.degree(v) != 1:
-                P.add_node((u, v))
-    for (u1, v1), (u2, v2) in itertools.combinations(P.nodes(), 2):
-        both_edge = G.has_edge(u1, u2) and H.has_edge(v1, v2)
-        neither_edge = not G.has_edge(u1, u2) and not H.has_edge(v1, v2)
-        # Effectively: not (G.has_edge(u1, u2) xor H.has_edge(v1, v2))
-        if u1 != u2 and v1 != v2 and (both_edge or neither_edge):
-            P.add_edge((u1, v1), (u2, v2))
-    cliques = nx.find_cliques(P)
+    for g1_node, g2_node in itertools.product(graph1, graph2):
+        if all(graph1.node[g1_node][attr] == graph2.node[g2_node][attr] for attr in attributes):
+            if graph1.degree(g1_node) != 1 and graph2.degree(g2_node) != 1:
+                product.add_node((g1_node, g2_node))
+    for (g1_node1, g2_node1), (g1_node2, g2_node2) in itertools.combinations(product.nodes(), 2):
+        both_edge = graph1.has_edge(g1_node1, g1_node2) and graph2.has_edge(g2_node1, g2_node2)
+        neither_edge = not graph1.has_edge(g1_node1, g1_node2) and not graph2.has_edge(g2_node1, g2_node2)
+        # Effectively: not (graph1.has_edge(g1_node1, g1_node2) xor graph2.has_edge(g2_node1, g2_node2))
+        if g1_node1 != g1_node2 and g2_node1 != g2_node2 and (both_edge or neither_edge):
+            product.add_edge((g1_node1, g2_node1), (g1_node2, g2_node2))
+    cliques = nx.find_cliques(product)
     largest = maxes(cliques, key=len)
 
-    # Now, for every MCS we found, look at the nodes of degree 1. The
+    # Now, for every MCS we graph2, look at the nodes of degree 1. The
```

Comment:
"very MCS we graph2", overkill search and replace?

Reply:
Very much so